scala/lambdas/ingest-validate-generic-ingest-inputs/src/main/scala/uk/gov/nationalarchives/ingestvalidategenericingestinputs/MetadataJsonValueValidator.scala

```diff
@@ -0,0 +1,392 @@
+package uk.gov.nationalarchives.ingestvalidategenericingestinputs
+
+import cats.data.*
+import cats.effect.IO
+import cats.implicits.*
+import org.scanamo.generic.semiauto.*
+import ujson.*
+import uk.gov.nationalarchives.DAS3Client
+import uk.gov.nationalarchives.dynamoformatters.DynamoFormatters.*
+import uk.gov.nationalarchives.ingestvalidategenericingestinputs.EntryType.*
+import uk.gov.nationalarchives.ingestvalidategenericingestinputs.EntryValidationError.*
+import uk.gov.nationalarchives.ingestvalidategenericingestinputs.MetadataJsonValueValidator.*
+import uk.gov.nationalarchives.ingestvalidategenericingestinputs.ValidatedUtils.{ValidatedEntry, given}
+
+import java.net.URI
+import java.util.UUID
+import scala.annotation.tailrec
+import scala.collection.immutable.Map
+import scala.util.{Failure, Success, Try}
+
+class MetadataJsonValueValidator {
+  private val parentId = "parentId"
+  private val series = "series"
+
+  def checkFileIsInCorrectS3Location(s3Client: DAS3Client[IO], fileEntries: List[ValidatedEntry]): IO[List[ValidatedNel[ValidationError, Value]]] =
+    fileEntries.map { fileEntry =>
+      val locationNel = fileEntry(location)
+      locationNel match {
+        case Validated.Valid(locationObject) =>
+          val fileLocation = locationObject.str
+          val potentialFileUri = Try(URI.create(fileLocation))
+          potentialFileUri match {
+            case Success(fileUri) if fileUri.getHost != null =>
+              s3Client
+                .headObject(fileUri.getHost, fileUri.getPath.drop(1))
+                .redeem(
+                  exception => NoFileAtS3LocationError(fileLocation, exception.getMessage).invalidNel[Value],
+                  headResponse =>
+                    val statusCode = headResponse.sdkHttpResponse().statusCode()
+                    if statusCode == 200 then locationNel
+                    else NoFileAtS3LocationError(fileLocation, s"Head Object request returned a Status code of $statusCode").invalidNel[Value]
+                )
+            case Success(_) => IO.pure(UriIsNotValid(fileLocation, s"'$location' could not be transformed into a URI").invalidNel[Value])
+            case Failure(e) => IO.pure(UriIsNotValid(fileLocation, e.getMessage).invalidNel[Value])
+          }
+        case locationInvalidNel => IO.pure(locationInvalidNel)
+      }
+    }.sequence
+
+  def checkFileNamesHaveExtensions(fileEntries: List[ValidatedEntry]): List[ValidatedEntry] =
```

Comment:
Ah, yes. I decided to only use this for validating metadata files (since they should end with `.json`) so in the lambda, I'm calling it with only metadata entries. I'll change the name of this to "checkMetadataFileNamesHaveExtensions"

Reply:
Ah, I saw this comment after I left [my comment here](https://github.com/nationalarchives/dr2-ingest/pull/242#discussion_r1765391783); I think this one will cause an issue if it's missing.